spyderlib/utils/programs.py

```diff
@@ -55,13 +55,79 @@ def find_program(basename):
             return path
 
 
-def run_program(name, args=[], cwd=None):
-    """Run program in a separate process"""
-    assert isinstance(args, (tuple, list))
-    path = find_program(name)
-    if not path:
-        raise RuntimeError("Program %s was not found" % name)
-    subprocess.Popen([path]+args, cwd=cwd)
+def alter_subprocess_kwargs_by_platform(**kwargs):
+    """Given a dict, populate kwargs to create a generally
+    useful default setup for running subprocess processes
+    on different platforms. For example, `close_fds` is
+    set on posix and creation of a new console window is
+    disabled on Windows.
+
+    This function will alter the given kwargs and return
+    the modified dict."""
+    kwargs.setdefault('close_fds', os.name == 'posix')
+    if os.name == 'nt':
+        CONSOLE_CREATION_FLAGS = 0  # Default value
+        # Add more here..
+        CREATE_NO_WINDOW = 0x08000000
+        # We "or" them together
+        CONSOLE_CREATION_FLAGS |= CREATE_NO_WINDOW
+        kwargs.setdefault('creationflags', CONSOLE_CREATION_FLAGS)
+    return kwargs
+
+
+def run_shell_command(cmdstr, **subprocess_kwargs):
+    """ Execute the given shell command. Note that *args and
+    **kwargs will be passed to the subprocess call. 
+    
+    If 'shell' is given in subprocess_kwargs it must be True.
+    If 'executable' is not given in subprocess_kwargs, it will
+    be set to the value of the SHELL environment variable.
+    
+    :str cmdstr: The string run as a shell command.
+    :subprocess_kwargs: These will be passed to subprocess.Popen."""
+    kwargs = subprocess_kwargs
+    if 'shell' in kwargs:
+        assert kwargs['shell']==True, (
+                'The "shell" kwarg may be omitted, but if '
+                'provided it must be True.')
+    else:
+        kwargs['shell'] = True
+
+    if 'executable' not in kwargs:
+        kwargs['executable'] = os.getenv('SHELL')
+
+    kwargs = alter_subprocess_kwargs_by_platform(**kwargs)
+    return subprocess.Popen(cmdstr, **kwargs)
+
+
+def run_program(program, args=[], **subprocess_kwargs):

```

Comment:
@Nodd Will change thanks. Originally, the reason I chose `args=[]` rather than `args=None` was because there were no doctoring hints as to the required type of args, and having the default be an empty list instead of None will give the caller a hint (e.g. if they use jedi for completion etc.).  Since that first draft I have added sphinx types in the docstring, so editors will show the docstring types and we don't need to hint it with the default type.  I will change it to `None`.